src/coreclr/jit/lower.cpp

```diff
@@ -7976,6 +7976,71 @@ void Lowering::TransformUnusedIndirection(GenTreeIndir* ind, Compiler* comp, Bas
     }
 }
 
+//------------------------------------------------------------------------
+// LowerLclHeap: a common logic to lower LCLHEAP.
+//
+// Arguments:
+//    blkNode - the LCLHEAP node we are lowering.
+//
+void Lowering::LowerLclHeap(GenTree* node)
+{
+    assert(node->OperIs(GT_LCLHEAP));
+
+#if defined(TARGET_XARCH)
+    if (node->gtGetOp1()->IsCnsIntOrI())
+    {
+        GenTreeIntCon* sizeNode = node->gtGetOp1()->AsIntCon();
+        assert((unsigned)sizeNode->IconValue() <= UINT_MAX);
+        sizeNode->SetIconValue(ALIGN_UP(sizeNode->IconValue(), STACK_ALIGN));
+        if (comp->info.compInitMem)
+        {
+            ssize_t size = node->gtGetOp1()->AsIntCon()->IconValue();
+            if ((unsigned)size > UINT_MAX)
+            {
+                BADCODE("Localloc size is negative or too big");
+            }
+
+            LIR::Use use;
+            if (BlockRange().TryGetUse(node, &use))
+            {
+                // Replace with null for LCLHEAP(0)
+                if (size == 0)
+                {
+                    GenTree* nullNode = comp->gtNewIconNode(0, node->TypeGet());
+                    use.ReplaceWith(nullNode);
+                    BlockRange().Remove(sizeNode);
+                    BlockRange().Remove(node);
+                    BlockRange().InsertAfter(node, nullNode);
+                    return;
+                }
+
+                // Emit STORE_BLK to zero it
+                //
+                //  *  STORE_BLK struct<size> (init) (Unroll)
+                //  +--*  LCL_VAR   long   V01
+                //  \--*  CNS_INT   int    0
+                //
+                GenTree*    heapLcl  = comp->gtNewLclvNode(use.ReplaceWithLclVar(comp), TYP_I_IMPL);
+                GenTree*    zero     = comp->gtNewIconNode(0);
+                GenTreeBlk* storeBlk = new (comp, GT_STORE_BLK)
+                    GenTreeBlk(GT_STORE_BLK, TYP_STRUCT, heapLcl, zero, comp->typGetBlkLayout((unsigned)size));
+                storeBlk->gtFlags |= (GTF_BLK_UNALIGNED | GTF_ASG | GTF_EXCEPT | GTF_GLOB_REF);
+                BlockRange().InsertAfter(use.Def(), heapLcl, zero, storeBlk);
+                LowerNode(storeBlk);
+                node->gtFlags |= GTF_LCLHEAP_ZEROED;
```

Comment:
The flag is only used in asserts AFAICT, do we need it?

(It can be a `GTF_DEBUG_` flag if it is for asserts only)

Reply:
Yeah mostly for myself to validate on CI